subsys/logging/log_core.c

```diff
@@ -22,12 +22,43 @@
 LOG_BACKEND_UART_DEFINE(log_backend_uart);
 #endif
 
+#define LOG_STATE_UNINITIALIZED		0
+#define LOG_STATE_INIT_IN_PROGRESS	1
+#define LOG_STATE_INITIALIZED		2
+
 static struct log_list_t list;
 static bool panic_mode;
-static bool initialized;
-
+static atomic_t log_state;
 static timestamp_get_t timestamp_func;
 
+static int init(void);
+static bool cond_init(void);
+
+/** @brief Macro for implicit logger initialization.
+ *
+ * If logger is used before being explicitly initialized it must be
+ * initialized implicitly and preemption must be considered. First flag
+ * is checked if log is initialized. Action is performed only if logger is not
+ * yet initialized. This initial check is performed in macro for performance
+ * reasons.
+ */
+#define _COND_INIT(_return)					\
+	do {							\
+		if (log_state < LOG_STATE_INITIALIZED) {	\
+			 if (!cond_init()) {			\
+				 _return;			\
+			 }					\
+		}						\
+	} while (0)
+
+#define LOG_COND_INIT()	_COND_INIT(return)
+#define LOG_COND_INIT_PRINTK()	_COND_INIT(return 0)
```

Comment:
Hiding return statement in macro is not a good idea. Please consider using anonymous functions to check the state and initialize logger.